pipeline/models.py

```diff
@@ -23,47 +46,198 @@ def at_least_one_output(cls, outputs: Dict[str, str]) -> Dict[str, str]:
+                    raise ValueError(f"Output path {filename} is not permitted: {e}")
+
+        return outputs
+
 
 class Command(BaseModel):
-    run: str  # original string
-    name: str
-    version: str
-    args: str
+    raw: str  # original string
 
     class Config:
         # this makes Command hashable, which for some reason due to the
         # Action.parse_run_string works, pydantic requires.
         frozen = True
 
+    @property
+    def args(self) -> str:
+        return " ".join(self.parts[1:])
+
+    @property
+    def name(self) -> str:
+        # parts[0] with version split off
+        return self.parts[0].split(":")[0]
+
+    @property
+    def parts(self) -> List[str]:
+        return shlex.split(self.raw)
+
+    @property
+    def version(self) -> str:
+        # parts[0] with name split off
+        return self.parts[0].split(":")[1]
+
 
 class Action(BaseModel):
+    config: Optional[Dict[Any, Any]] = None
     run: Command
     needs: List[str] = []
     outputs: Outputs
+    dummy_data_file: Optional[pathlib.Path]
 
     @validator("run", pre=True)
     def parse_run_string(cls, run: str) -> Command:
         parts = shlex.split(run)
+
         name, _, version = parts[0].partition(":")
-        args = " ".join(parts[1:])
-        return Command(
-            run=run,
-            name=name,
-            version=version,
-            args=args,
-        )
+        if not version:
+            raise ValueError(
+                f"{name} must have a version specified (e.g. {name}:0.5.2)",
+            )
+
+        return Command(raw=run)
+
+
+class PartiallyValidatedPipeline(TypedDict):
+    """
+    A custom type to type-check the values in "post" root validators
+
+    A root_validator with pre=False (or no kwargs) runs after the values have
+    been ingested already, and the `values` arg is a dictionary of model types.
+
+    Note: This is defined here so we don't have to deal with forward reference
+    types.
+    """
+
+    version: float
+    expectations: Expectations
+    actions: Dict[str, Action]
 
 
 class Pipeline(BaseModel):
-    version: Union[str, float]
+    version: float
     expectations: Expectations
     actions: Dict[str, Action]
 
+    @property
+    def all_actions(self) -> List[str]:
+        """
+        Get all actions for this Pipeline instance
+
+        We ignore any manually defined run_all action (in later project
+        versions this will be an error). We use a list comprehension rather
+        than set operators as previously so we preserve the original order.
+        """
+        return [action for action in self.actions.keys() if action != RUN_ALL_COMMAND]
+
+    @root_validator()
+    def validate_actions(
+        cls, values: PartiallyValidatedPipeline
+    ) -> PartiallyValidatedPipeline:
+        # TODO: move to Action when we move name onto it
+        for action_id, config in values.get("actions", {}).items():
+            validators = {
```

Comment:
This dict shouldn't be defined inside the loop, as its constant.

It could be a static class attribute?

Reply:
Opted to move it outside of the loop where I meant to put it in the first place!